notebooks/ch11/gradient_stochastic.md

```diff
@@ -0,0 +1,331 @@
+
+### Visualizing Stochastic Gradient Descent
+
+Below are visual examples of loss minimization in batch gradient descent and stochastic gradient descent.
+
+![](gd.png)
+![](sgd.png)
+
+At each iteration of batch gradient descent, we move in the direction of the true gradient of the loss function, which is depicted by the ellipses. On the other hand, each iteration of stochastic gradient descent may not lead us in the direction of the true gradient; however, the $\hat\theta$ parameters eventually converge to the minima of the loss function. Although stochastic gradient descent typically takes more iterations to converge than batch gradient descent, it often converges more quickly because it spends significantly less time computing the update at each iteration.
+
+
+```python
+##### 4 TEMPORARY CELLS BELOW TO TEST OUT %%TIME FOR BATCH GRADIENT DESCENT.
+```
+
+
+```python
+def minimize(loss_fn, grad_loss_fn, dataset, alpha=0.2, progress=True):
+    '''
+    Uses gradient descent to minimize loss_fn. Returns the minimizing value of
+    theta once theta changes less than 0.001 between iterations.
+    '''
+    theta = 0
+    while True:
+        if progress:
+            print(f'theta: {theta:.2f} | loss: {loss_fn(theta, dataset):.2f}')
+        gradient = grad_loss_fn(theta, dataset)
+        new_theta = theta - alpha * gradient
+        
+        if abs(new_theta - theta) < 0.001:
+            return new_theta
+        
+        theta = new_theta
+```
+
+
+```python
+# defining sets of sample_data and entire_data to test out gd, sgd, mbgd
+sample_data = np.array([12.1, 12.8, 14.9, 16.3, 17.2])
+entire_data = np.array(tips['pcttip'])
+```
+
+
+```python
+# This yields the correct theta that the algorithm should converge to.
+np.mean(entire_data)
+```
+
+
+
+
+    16.08025817225047
+
+
+
+
+```python
+%%time
+minimize(mse_loss, grad_mse_loss, entire_data)
+```
+
+    theta: 0.00 | loss: 295.72
+    theta: 6.43 | loss: 130.23
+    theta: 10.29 | loss: 70.66
+    theta: 12.61 | loss: 49.21
+    theta: 14.00 | loss: 41.49
+    theta: 14.83 | loss: 38.71
+    theta: 15.33 | loss: 37.71
+    theta: 15.63 | loss: 37.35
+    theta: 15.81 | loss: 37.22
+    theta: 15.92 | loss: 37.17
+    theta: 15.98 | loss: 37.15
+    theta: 16.02 | loss: 37.15
+    theta: 16.05 | loss: 37.15
+    theta: 16.06 | loss: 37.15
+    theta: 16.07 | loss: 37.15
+    theta: 16.07 | loss: 37.15
+    theta: 16.08 | loss: 37.15
+    theta: 16.08 | loss: 37.15
+    theta: 16.08 | loss: 37.15
+    CPU times: user 15.6 ms, sys: 0 ns, total: 15.6 ms
+    Wall time: 1.75 ms
+
+
+
+
+
+    16.07927830605656
+
+
+
+### Defining a Function for Stochastic Gradient Descent
+
+As we previously did for batch gradient descent, we would like to define a function that computes the stochastic gradient descent of the loss function. It will be similar to our `minimize` function, but we will need to implement the random selection of one observation at each iteration.
```

Comment:
We don't need to say "would like" since we actually define it. Let's just keep this paragraph in present tense.

The last sentence shouldn't have a comma before the "but" since the first part of the sentence can't become a standalone sentence.